File: backend/domain/policyruleclassification/policyruleclassification.go
// SPDX-FileCopyrightText: 2025 Mercedes-Benz Group AG and Mercedes-Benz AG
//
// SPDX-License-Identifier: Apache-2.0

package policyruleclassification

import (
	"time"

	"github.com/eclipse-disuko/disuko/domain"
)

type RuleStatus string
type ClassificationKey string

const (
	RuleStatusAllowed   RuleStatus = "allowed"
	RuleStatusWarned    RuleStatus = "warned"
	RuleStatusDenied    RuleStatus = "denied"
	RuleStatusForbidden RuleStatus = "forbidden"
)

type PolicyRuleClassification struct {
	domain.RootEntity `bson:",inline"`
	Name              string                           `json:"name"`
	Rules             map[ClassificationKey]RuleStatus `json:"rules"`
}

func (p *PolicyRuleClassification) Update(name string, rules map[ClassificationKey]RuleStatus) {
	p.Name = name
	p.Rules = rules
	p.Updated = time.Now()
}

File: frontend/libs/portal/model/PolicyRuleClassification.ts
// SPDX-FileCopyrightText: 2025 Mercedes-Benz Group AG and Mercedes-Benz AG
//
// SPDX-License-Identifier: Apache-2.0

import {BaseDto} from '@disclosure-portal/model/BaseClass';

export type RuleStatus = 'allowed' | 'warned' | 'denied' | 'forbidden';

export const DEFAULT_CLASSIFICATION_NAMES: ReadonlyArray<string> = [
  'Permissive',
  'Copyleft (weak)',
  'Copyleft (strong)',
  'Copyleft (network protective)',
  'Anti-Tivoization',
  'Severe patent retaliation',
  'Doing Business with US',
  'Unclear or Ambiguous',
];

export interface ClassificationInfo {
  key: string;
  name: string;
}

export interface MatrixResponseDto {
  classifications: ClassificationInfo[];
  useCases: PolicyRuleClassificationDto[];
}

export interface PolicyRuleClassificationRequestDto {
  name: string;
  rules: Record<string, RuleStatus>;
}

export class PolicyRuleClassificationDto extends BaseDto {
  public name = '';
  public rules: Record<string, RuleStatus> = {};

  public constructor(dto?: PolicyRuleClassificationDto | null) {
    super(dto);
  }
}
